File: server/controllers/shortlistInfoController.js
const ShortlistInfoModel = require("../models/shortlistInfoModel");
const xlsx = require("xlsx");
const pool = require("../config/db");
const path = require('path');
const fs = require('fs');

const shortlistInfoController = {
    // Fetch all shortlist batch names for a specific year
    getShortlistNames: async (req, res) => {
        const { year } = req.query;
        try {
            const names = await ShortlistInfoModel.getAllShortlistNames(year);
            res.json(names);
        } catch (error) {
            console.error("getShortlistNames Error:", error);
            res.status(500).json({ message: "Error fetching shortlist names", error: error.message });
        }
    },

    // Fetch only non-frozen shortlist batch names for a specific year
    getNonFrozenShortlistNames: async (req, res) => {
        const { year } = req.query;
        try {
            const nonFrozenNames = await ShortlistInfoModel.getNonFrozenShortlistNames(year);
            res.json(nonFrozenNames);
        } catch (error) {
            console.error("getNonFrozenShortlistNames Error:", error);
            res.status(500).json({ message: "Error fetching non-frozen shortlist names", error: error.message });
        }
    },

    // Fetch detailed shortlist batch info
    getShortlistDetails: async (req, res) => {
        const { shortlistName } = req.params;
        const { year } = req.query;
        try {
            const info = await ShortlistInfoModel.getShortlistInfo(shortlistName, year);
            if (!info) return res.status(404).json({ message: "Shortlist not found" });
            res.json(info);
        } catch (error) {
            console.error(`getShortlistDetails Error [${shortlistName}]:`, error);
            res.status(500).json({ message: "Error fetching shortlist info", error: error.message });
        }
    },

    // Fetch total applicant and shortlisted counts
    getCounts: async (req, res) => {
        const { year } = req.query;
        try {
            const totalApplicants = await ShortlistInfoModel.getTotalApplicantCount(year);
            const totalShortlisted = await ShortlistInfoModel.getTotalShortlistedCount(year);
            res.json({ totalApplicants, totalShortlisted });
        } catch (error) {
            console.error("getCounts Error:", error);
            res.status(500).json({ message: "Error fetching counts", error: error.message });
        }
    },

    /**
     * 🔥 Main Freeze Logic
     * 1. Auto-updates single-medium school students.
     * 2. Detects multi-medium school conflicts.
     * 3. Freezes ONLY if zero conflicts remain.
     */
    freezeShortlist: async (req, res) => {
        const { shortlistBatchId, filterMediums } = req.body;

        if (!shortlistBatchId) return res.status(400).json({ message: "Batch ID required" });
        if (!filterMediums || filterMediums.length === 0) {
            return res.status(400).json({ message: "Select at least one medium" });
        }

        try {
            // STEP 1: Auto-update students in schools with ONLY ONE matching medium
            await ShortlistInfoModel.autoUpdateSingleMediumStudents(shortlistBatchId, filterMediums);

            // STEP 2: Fetch the "Others" (Students in multi-medium schools requiring manual review)
            const invalidStudents = await ShortlistInfoModel.getInvalidMediumStudents(shortlistBatchId, filterMediums);

            // STEP 3: If there are ANY remaining conflicts, STOP and return them to the UI
            if (invalidStudents.length > 0) {
                return res.status(400).json({
                    requiresCorrection: true,
                    message: `${invalidStudents.length} students require manual medium selection (Multi-medium schools detected).`,
                    students: invalidStudents
                });
            }

            // STEP 4: If invalidStudents is 0, freeze the batch
            const success = await ShortlistInfoModel.freezeShortlist(shortlistBatchId);
            if (success) {
                res.json({ message: "Shortlist filtered and frozen successfully" });
            } else {
                res.status(404).json({ message: "Shortlist not found or already frozen" });
            }
        } catch (error) {
            console.error(`freezeShortlist Error:`, error);
            res.status(500).json({ message: "Error during freeze process", error: error.message });
        }
    },

    // Handle manual decisions submitted from the correction table
    bulkUpdateMediums: async (req, res) => {
        const { updates, batchId } = req.body;
        if (!updates || !batchId) return res.status(400).json({ message: "Missing data" });
        try {
            const success = await ShortlistInfoModel.bulkUpdateMediumsAndStatus(updates, batchId);
            if (success) {
                res.json({ message: "Medium decisions updated successfully" });
            }
        } catch (error) {
            console.error("bulkUpdateMediums Error:", error);
            res.status(500).json({ message: "Failed to update student data", error: error.message });
        }
    },

    // Delete a shortlist batch
    deleteShortlist: async (req, res) => {
        const { shortlistBatchId } = req.body;
        const { year } = req.query;
        try {
            const success = await ShortlistInfoModel.deleteShortlist(shortlistBatchId, year);
            if (success) res.json({ message: "Shortlist deleted successfully" });
            else res.status(404).json({ message: "Shortlist not found" });
        } catch (error) {
            console.error(`deleteShortlist Error:`, error);
            res.status(500).json({ message: "Error deleting shortlist", error: error.message });
        }
    },

    // Get shortlisted applicants for display
    getShortlistedApplicantsForShow: async (req, res) => {
        const { shortlistName } = req.params;
        const { year } = req.query;
        try {
            const info = await ShortlistInfoModel.getShortlistInfo(shortlistName, year);
            if (!info) return res.status(404).json({ message: "Shortlist not found" });
            const data = await ShortlistInfoModel.getShortlistedApplicantsForShow(info.id, year);
            res.json({ name: info.name, data: data });
        } catch (error) {
            console.error(`getShortlistedApplicantsForShow Error:`, error);
            res.status(500).json({ message: "Error fetching display data", error: error.message });
        }
    },

    // Generate Excel and send for download
    getShortlistedApplicantsForDownload: async (req, res) => {
        const { shortlistName } = req.params;
        const { year: queryYear } = req.query;

        try {
            const year = queryYear;
            const shortlistInfo = await ShortlistInfoModel.getShortlistInfo(shortlistName, year);
            if (!shortlistInfo) return res.status(404).json({ message: "Shortlist not found" });

            // Check if data exists for download
            const totalStudentsRes = await pool.query(
                `SELECT COUNT(*) AS total_students 
                 FROM pp.applicant_shortlist_info 
                 WHERE shortlist_batch_id = $1 AND shortlisted_yn = 'Y'`,
                [shortlistInfo.id]
            );
            
            if (parseInt(totalStudentsRes.rows[0].total_students) === 0) {
                return res.status(200).json({ status: "no_data", message: "No shortlisted students found." });
            }

            let applicants = await ShortlistInfoModel.getShortlistedApplicantsForDownload(shortlistInfo.id, year);

            // Add Serial Number
            const formattedData = applicants.map((app, i) => ({ "S. No.": i + 1, ...app }));

            const fileName = `${shortlistName}_Applicants.xlsx`;
            const worksheet = xlsx.utils.json_to_sheet(formattedData);
            const workbook = xlsx.utils.book_new();
            xlsx.utils.book_append_sheet(workbook, worksheet, "Applicants");

            const FILE_DIR = path.join(process.env.FILE_STORAGE_PATH, 'generated-shortlist-data');
            if (!fs.existsSync(FILE_DIR)) {
                fs.mkdirSync(FILE_DIR, { recursive: true });
            }

            const localSavePath = path.join(FILE_DIR, fileName);
            xlsx.writeFile(workbook, localSavePath);

            const buffer = xlsx.write(workbook, { type: 'buffer', bookType: 'xlsx' });
            res.setHeader('Content-Disposition', `attachment; filename="${fileName}"`);
            res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
            return res.status(200).send(buffer);
        } catch (error) {
            console.error(`Download Error:`, error);
            res.status(500).json({ message: "Error generating download", error: error.message });
        }
    },

    // Reset medium filtering
    resetMediums: async (req, res) => {
        const { shortlistBatchId } = req.body;
        try {
            const success = await ShortlistInfoModel.resetMediumFiltering(shortlistBatchId);
            if (success) {
                res.json({ message: "Medium filtering reset successfully." });
            } else {
                res.status(400).json({ message: "Reset failed. Batch may be frozen." });
            }
        } catch (error) {
            console.error(`resetMediums Error:`, error);
            res.status(500).json({ message: "Error resetting filtering", error: error.message });
        }
    },
};

module.exports = shortlistInfoController;

File: java-backend/imas-backend/src/main/java/com/rcf/imas/modules/shortlist/web/ShortlistInfoController.java
package com.rcf.imas.modules.shortlist.web;

import com.rcf.imas.modules.shortlist.persistence.ShortlistReadRepository;
import com.rcf.imas.modules.shortlist.persistence.ShortlistWriteRepository;
import com.rcf.imas.modules.shortlist.service.XlsxSupport;
import com.rcf.imas.platform.error.ApiException;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/shortlist-info")
@PreAuthorize("hasRole('ADMIN')")
class ShortlistInfoController {

    private static final List<String> DOWNLOAD_HEADERS = List.of(
        "S. No.", "NMMS Registration No", "Student Name", "Contact No 1", "Current School Name",
        "Medium", "District", "Block", "GMAT Score", "SAT Score");

    private final ShortlistReadRepository reads;
    private final ShortlistWriteRepository writes;
    private final XlsxSupport xlsx;

    ShortlistInfoController(ShortlistReadRepository reads, ShortlistWriteRepository writes, XlsxSupport xlsx) {
        this.reads = reads;
        this.writes = writes;
        this.xlsx = xlsx;
    }

    @GetMapping("/names")
    public List<String> names(@RequestParam(required = false) String year) { return reads.shortlistNames(year); }

    @GetMapping("/non-frozen-names")
    public List<Map<String, Object>> nonFrozenNames(@RequestParam(required = false) String year) { return reads.nonFrozenNames(year); }

    @GetMapping("/counts")
    public Map<String, Object> counts(@RequestParam(required = false) String year) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("totalApplicants", reads.totalApplicantCount(year));
        m.put("totalShortlisted", reads.totalShortlistedCount(year));
        return m;
    }

    @GetMapping("/show-data/{shortlistName}")
    public Map<String, Object> showData(@PathVariable String shortlistName, @RequestParam(required = false) String year) {
        Map<String, Object> info = reads.shortlistInfo(shortlistName, year);
        if (info == null) throw ApiException.message(404, "Shortlist not found");
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("name", info.get("name"));
        out.put("data", reads.showData((String) info.get("id")));
        return out;
    }

    @GetMapping("/download-data/{shortlistName}")
    public Object downloadData(@PathVariable String shortlistName, @RequestParam(required = false) String year) {
        Map<String, Object> info = reads.shortlistInfo(shortlistName, year);
        if (info == null) throw ApiException.message(404, "Shortlist not found");
        String id = (String) info.get("id");
        if (reads.shortlistedCountInBatch(id) == 0) {
            return Map.of("status", "no_data", "message", "No shortlisted students found.");
        }
        List<Map<String, Object>> rows = reads.downloadRows(id);
        List<Map<String, Object>> withSno = new java.util.ArrayList<>();
        for (int i = 0; i < rows.size(); i++) {
            Map<String, Object> r = new LinkedHashMap<>();
            r.put("S. No.", i + 1);
            r.putAll(rows.get(i));
            withSno.add(r);
        }
        byte[] bytes = xlsx.build("Applicants", DOWNLOAD_HEADERS, withSno);
        return org.springframework.http.ResponseEntity.ok()
            .header("Content-Disposition", "attachment; filename=\"" + shortlistName + "_Applicants.xlsx\"")
            .header("Content-Type", "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")
            .body(bytes);
    }

    @GetMapping("/{shortlistName}")
    public Map<String, Object> detail(@PathVariable String shortlistName, @RequestParam(required = false) String year) {
        Map<String, Object> info = reads.shortlistInfo(shortlistName, year);
        if (info == null) throw ApiException.message(404, "Shortlist not found");
        return info;
    }

    @PostMapping("/freeze")
    public Object freeze(@RequestBody(required = false) Map<String, Object> body) {
        Map<String, Object> b = body == null ? Map.of() : body;
        Object batchIdRaw = b.get("shortlistBatchId");
        if (batchIdRaw == null || String.valueOf(batchIdRaw).isBlank())
            throw ApiException.message(400, "Batch ID required");
        @SuppressWarnings("unchecked")
        List<Object> filterMediums = b.get("filterMediums") instanceof List<?> l ? (List<Object>) l : List.of();
        if (filterMediums.isEmpty()) throw ApiException.message(400, "Select at least one medium");

        String batchId = String.valueOf(batchIdRaw);
        List<String> allowed = filterMediums.stream().map(String::valueOf).toList();

        writes.autoUpdateSingleMediumStudents(batchId);
        List<Map<String, Object>> invalid = reads.invalidMediumStudents(batchId, allowed);
        if (!invalid.isEmpty()) {
            Map<String, Object> out = new LinkedHashMap<>();
            out.put("requiresCorrection", true);
            out.put("message", invalid.size() + " students require manual medium selection (Multi-medium schools detected).");
            out.put("students", invalid);
            // 400 body carries no error/message-key envelope — return a ResponseEntity directly
            return org.springframework.http.ResponseEntity.status(400).body(out);
        }
        if (writes.freezeShortlist(batchId)) {
            return Map.of("message", "Shortlist filtered and frozen successfully");
        }
        return org.springframework.http.ResponseEntity.status(404).body(Map.of("message", "Shortlist not found or already frozen"));
    }

    @PostMapping("/bulk-update-mediums")
    @SuppressWarnings("unchecked")
    public Map<String, Object> bulkUpdateMediums(@RequestBody(required = false) Map<String, Object> body) {
        Map<String, Object> b = body == null ? Map.of() : body;
        Object updatesRaw = b.get("updates");
        Object batchIdRaw = b.get("batchId");
        if (!(updatesRaw instanceof List<?>) || batchIdRaw == null || String.valueOf(batchIdRaw).isBlank())
            throw ApiException.message(400, "Missing data");
        List<Map<String, Object>> updates = (List<Map<String, Object>>) updatesRaw;
        writes.bulkUpdateMediumsAndStatus(updates, String.valueOf(batchIdRaw));
        return Map.of("message", "Medium decisions updated successfully");
    }

    @PostMapping("/reset-mediums")
    public Object resetMediums(@RequestBody(required = false) Map<String, Object> body) {
        Map<String, Object> b = body == null ? Map.of() : body;
        String batchId = b.get("shortlistBatchId") == null ? null : String.valueOf(b.get("shortlistBatchId"));
        if (batchId != null && writes.resetMediumFiltering(batchId)) {
            return Map.of("message", "Medium filtering reset successfully.");
        }
        return org.springframework.http.ResponseEntity.status(400).body(Map.of("message", "Reset failed. Batch may be frozen."));
    }

    @DeleteMapping("/delete")
    public Object deleteShortlist(@RequestBody(required = false) Map<String, Object> body,
                                  @RequestParam(required = false) String year) {
        Map<String, Object> b = body == null ? Map.of() : body;
        String batchId = b.get("shortlistBatchId") == null ? null : String.valueOf(b.get("shortlistBatchId"));
        if (batchId != null && writes.deleteShortlist(batchId)) {
            return Map.of("message", "Shortlist deleted successfully");
        }
        return org.springframework.http.ResponseEntity.status(404).body(Map.of("message", "Shortlist not found"));
    }
}
